usb/util.py

```diff
@@ -241,6 +242,53 @@ def dispose_resources(device):
     """
     device._ctx.dispose(device)
 
+def get_langids(dev):
+    r"""Retrieve the list of supported Language IDs from the device.
+
+    USB LANGIDs are 16-bit integers familiar to Windows developers, where
+    for example instead of en-US you say 0x0409. See the file USB_LANGIDS.pdf
+    somewhere on the usb.org site for a list, which does not claim to be
+    complete. It requires "system software must allow the enumeration and
+    selection of LANGIDs that are not currently on this list." It also requires
+    "system software should never request a LANGID not defined in the LANGID
+    code array (string index = 0) presented by a device." So it is good to
+    expose that array to client code.
+
+    The array is retrieved by asking for string descriptor zero, which is
+    never the index of a real string. The returned descriptor has bLength
+    and bDescriptorType bytes followed by pairs of bytes representing
+    little-endian LANGIDs. In USB 3.1 r1.0 section 9.6.9, a device is allowed
+    to omit all string descriptors, and to not return an array of LANGID codes
+    at all. In that case client code should refrain from requesting any strings.
+    That case is covered by returning an empty array here, as no client should
+    request any string with a LANGID that isn't in the array.
+    """
+    from usb.control import get_descriptor
+    buf = get_descriptor(
+                dev,
+                254,
+                DESC_TYPE_STRING,
+                0
+            )
+
+    langids = array.array('H')
+
+    if len(buf) < 4 or buf[0] < 4:
+        return langids
+
+    buf = buf[2:buf[0]]
+    buf.extend((0, 1)) # in proper endianness this should represent 256
+
+    if hexversion >= 0x03020000:
+        langids.frombytes(buf.tobytes())
+    else:
+        langids.fromstring(buf.tostring())
```

Comment:
get_descriptor returns an array.array('B') object, but LANGIDs are short ints and should be presented that way to the client. Original array.array seems not to have an easy cast to a new shape method as Python 3 memoryviews do. However, as I think more, there is no need for this to be fancy at all. LANGIDs could be returned as a plain Python tuple of ints. Simpler all around. Absent any objection, I'll change it to that.